### PR TITLE
adding git dhcpv4 client server yang model

### PR DESCRIPTION
**Pull request by SuryyaKrJana**
The DHCPv4 Client and Server yang models for States.
OTG Model for DHCPv4 client and server:  https://github.com/open-traffic-generator/models/pull/371

https://github.com/open-traffic-generator/models-yang/blob/dhcpv4-yang/artifacts/open-traffic-generator-dhcpv4client.txt
https://github.com/open-traffic-generator/models-yang/blob/dhcpv4-yang/artifacts/open-traffic-generator-dhcpv4server.txt

Redoc Reference:
 https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/open-traffic-generator/models/dhcpv4_support_review/artifacts/openapi.yaml&nocors#tag/Configuration

Path for fetching Client Interface information:
// Path from root: "/dhcpv4-clients/dhcpv4-client[name=dhpc4-client1]/state/
// Path from root: "/dhcpv4-clients/dhcpv4-client[name=dhpc4-client1]/state/interface/address"
// Path from root: "/dhcpv4-clients/dhcpv4-client[name=dhpc4-client1]/state/interface/prefix-length"
// Path from root: "/dhcpv4-clients/dhcpv4-client[name=dhpc4-client1]/state/interface/gateway_address"
// Path from root: "/dhcpv4-clients/dhcpv4-client[name=dhpc4-client1]/state/interface/lease_time"
// Path from root: "/dhcpv4-clients/dhcpv4-client[name=dhpc4-client1]/state/interface/renew_time"

### CODE REVIEW

File: models/dhcp/v4server/open-traffic-generator-dhcpv4server.yang [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+      type otg-types:counter64;
+      description
+        "Number of RELEASE messages received.";
+    }
+    leaf declines-received {
+      type otg-types:counter64;
+      description
+        "Number of RELEASE messages received.";
+    }
+  }
```
Comment (apratimmukherjee) [inline models/dhcp/v4server/open-traffic-generator-dhcpv4server.yang:125]: Their seems to be a DHCPINFORM pdu in https://datatracker.ietf.org/doc/html/draft-liu-dhc-dhcp-yang-model-07 . Is this not needed for OTG ?

File: models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      type otg-types:counter64;
+      description
+        "Number of RELEASE messages sent.";
+    }
+    leaf declines-sent {
+      type otg-types:counter64;
+      description
+        "Number of RELEASE messages sent.";
+    }
+  }
```
Comment (apratimmukherjee) [inline models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang:139]: Their seems to be a DHCPINFORM pdu in https://datatracker.ietf.org/doc/html/draft-liu-dhc-dhcp-yang-model-07 . Is this not needed for OTG ?

File: models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    }
+    leaf renew_time {
+      type uint32;
+      description
+        "Time in seconds until the DHCPv4 client starts renewing the lease.";
+    }
+    leaf rebind_time {
+      type uint32;
+      description
+        "Time in seconds until the DHCPv4 client starts rebinding.";
```
Comment (apratimmukherjee) [inline models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang:177]: If there is a choice of servers , might some sort of id/ip  of the DHCP server which provided this address be helpful. Or is the server normally a single one so this is not important ?

File: models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+      uses dhcp4client-address;
+    }
+  }
+
+  grouping dhcp4client-counters {
+    description
+      "Counters of the DHCPv4 Client.";
+
+    leaf sessions-up {
```
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang:94]: the stats publish in IxN is red/green nothing related to dhcp state m/c. I will removed.

File: models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang (hunk 2 above)
Comment (SuryyaKrJana) [reply models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang:139]: This comes as part of right-click option in IxN. We are not supporting for first option.

File: models/dhcp/v4server/open-traffic-generator-dhcpv4server.yang (hunk 1 above)
Comment (SuryyaKrJana) [reply models/dhcp/v4server/open-traffic-generator-dhcpv4server.yang:125]: This comes as part of right-click option in IxN. We are not supporting for first option.

File: requirements.txt [hunk 5]
```diff
@@ -1,2 +1,2 @@
-setuptools 
+setuptools
```
[the remark this answers is not in the document]
Comment (apratimmukherjee) [reply requirements.txt:1]: This is related to CI issue related to latest python. Should probably do this in separate PR linked to different issue like pipeline failure at head of main , merge to main and pull that into this branch.

File: models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang (hunk 3 above)
Comment (SuryyaKrJana) [reply models/dhcp/v4client/open-traffic-generator-dhcpv4client.yang:177]: server-id or server-ip: IxNetwork is not implemented this. First, we need to add this enhancement there,

**Review by apratimmukherjee — APPROVED** [review]
LGTM from model point of view.

**Review by dplore — APPROVED** [review]
at a glance, LGTM.  For our use case, validating DHCP relay on a DUT, biggest interest is triggering a client DISCOVER,  observing the server receive the relay'd request and the client receiving the OFFER and finally observing the lease on client and server.

I don't see data for dhcp options?  RFC 3046 option 82 might be nice to see.

**Comment by SuryyaKrJana** [comment]
> at a glance, LGTM. For our use case, validating DHCP relay on a DUT, biggest interest is triggering a client DISCOVER, observing the server receive the relay'd request and the client receiving the OFFER and finally observing the lease on client and server.

_Thanks for your feedback._ 

> I don't see data for dhcp options? RFC 3046 option 82 might be nice to see.
I need to work on further to support "RFC 3046 option 82". May be this can be done in next phase.

Are you looking for 'option 82' similar to 'host' under 'leases' in the server-side yang-model referred below? If you have any reference for open-model that will be great to be look at too.

[https://github.com/open-traffic-generator/models-yang/blob/dhcpv4-yang/artifacts/open-traffic-generator-dhcpv4server.txt](https://www.google.com/url?q=https://github.com/open-traffic-generator/models-yang/blob/dhcpv4-yang/artifacts/open-traffic-generator-dhcpv4server.txt&sa=D&source=buganizer&usg=AOvVaw3f0ujSdPabPVZIvasUT4d_)
module: open-traffic-generator-dhcpv4server
  +--rw dhcpv4-servers
     +--ro servers* [name]
        +--ro name     -> ../state/name
        +--ro state
           +--ro name?       string
           +--ro counters
           |  +--ro discovers-received?   otg-types:counter64
           |  +--ro offers-sent?          otg-types:counter64
           |  +--ro requests-received?    otg-types:counter64
           |  +--ro acks-sent?            otg-types:counter64
           |  +--ro nacks-sent?           otg-types:counter64
           |  +--ro releases-received?    otg-types:counter64
           |  +--ro declines-received?    otg-types:counter64
           +--ro leases
              +--ro leases-states
                 +--ro hosts* [address]
                    +--ro address    -> ../state/address
                    +--ro state
                       +--ro address?          otg-types:ipv4-address
                       +--ro valid-time?       uint32
                       +--ro preferred_time?   uint32
                       +--ro renew_time?       uint32
                       +--ro rebind_time?      uint32
                       +--ro client-id?        string
                       +--ro circuit-id?       string

**Pull request merged**